### ISSUE TITLE
onAdding has null value: autocomplete using a remote endpoint, 

### ISSUE THREAD

**Issue by magarbikram** [Gbuomprisco/ng2-tag-input#417]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request/question
```

**Current behavior**
1. autocomplete using a remote endpoint
2. onlyFromAutocomplete='false', so that user can input new tag as well

when user inputs a new tag, 

onAdding event will be called with 'undefined' input

**Expected behavior**

onAdding event should be called with value/display pair, as it does it onAdd event

**Minimal reproduction of the problem with instructions (if applicable)**

<tag-input [ngModel]="['@item']" [onAdding]='onAdding'>
       <tag-input-dropdown [autocompleteItems]="[{display: 'Item1', value: 0}]">
       </tag-input-dropdown>
</tag-input>

input 'xItem' and hit Enter

onAdding: method will be called with undefined value

**What is the motivation / use case for changing the behavior? (if applicable)**

1. I would like to intercept onAdding method, and if it is not in system, would like to add new tag in server and server will respond with a response 
2. I will return Observable<Response> for onAdding method, so that tag will have valid tag every time.

**What do you use to build your app? (SystemJS, Webpack, angular-cli, etc.). Please specify the version**

angular-cli

**Angular version:** 

4.0

**ng2-tag-input version:**

1.2.9

**Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

all

**Comment by Gbuomprisco** [comment]
Hi, please paste here your onAdding method

**Comment by magarbikram** [comment]
onAddingSinger = (singer: any): Observable<ArtistRole> => {
     ///other code goes here
}

<tag-input [(ngModel)]="singers"
                           theme='bootstrap' 
                           placeholder='Add more singer' 
                           secondaryPlaceholder='Add singer'
                           [onAdding]='onAddingSinger'
                           onlyFromAutocomplete='false'
                           identifyBy='artistRoleId'
                           displayBy='artistName'>
                    <tag-input-dropdown [autocompleteObservable]='requestAutocompleteItems'
                                        identifyBy='artistRoleId'
                                        displayBy='artistName'>
                        <ng-template let-item="item" let-index="index">
                          <img style="width:50px;height: 50px; padding:5px;" src="{{item.profilePicture?.imageUrl}}" 
                          alt="{{item.artistName}}">
                            {{ item.artistName }}
                        </ng-template>
                    </tag-input-dropdown>
                </tag-input>

**Comment by Gbuomprisco** [comment]
Can you paste the whole onAdding method?

**Comment by magarbikram** [comment]
Sure, here it is
i got singer undefined, in that particular case where the input doesn't match with auto-complete options

[Code]:

onAddingSinger = (singer: any): Observable<ArtistRole> => {
    if (singer && 'artistRoleId' in singer) {
      if (singer.artistRoleId > 0) {
        console.log('a new singer is added: ' + singer.artistName);
        return Observable.from([singer]);
      }
      else {
        if (confirm('Would you like to assign an artist in singer role: ' + singer.artistName)) {
          return this.assignArtistAsSingerService.assign(singer.artistId);
        } else {
          return Observable.from([]);
        }
      }
    }
    else if(singer) {
      console.log('a new singer should be added into the system: ' + singer.value);
      if (confirm('Would you like to add new singer in Nepalyrics: ' + singer.value)) {
        return this.addNewSingerService.add(singer.value);
      } else {
        return Observable.from([]);
      }
    }
    return Observable.from([singer]);
  }

**Comment by Gbuomprisco** [comment]
Hi @magarbikram,

I'm not sure why that happens. From my tests, if I select something from the autocomplete items, they are correctly passed to the onAdding fn, or if I enter them through the form. 

There are 2 things that catch my eye (but maybe i'm missing something myself...):
- You're passing an array, instead of the tag itself. Why not `return Observable.of(singer)?`
- I'd always return an Observable, but with a filter function so that it only gets added if the filter functions is truthy, which is how it's supposed to work. Instead, you're passing empty arrays. 

To summarize:
- don't pass arrays
- use a filter function to control what gets added, and what not
- always map to the tag itself

Example:

```javascript
public onAdding(tag): Observable<any> {
        const confirm = window.confirm('Do you really want to add this tag?');
        return Observable
            .of(undefined)
            .filter(() => confirm)
            .mapTo(tag);
}
```

Let me know if this is clearer, I understand the documentation is rather poor for these methods

**Comment by magarbikram** [comment]
Thanks for all the help so far, but I couldn't still run a very simple setup

**.html**
          
               <tag-input 
                    [(ngModel)]="tags"
                    [onAdding]='onAddingTag'>
                    <tag-input-dropdown [autocompleteItems]="['tag1', 'tag2', 'tag3','tag4']">
                    </tag-input-dropdown>
                </tag-input>

**.ts**
tags = ['tag1','tag2'];

public onAddingTag(tag):Observable<any>{
    debugger;   
    console.log('adding tag: ' + tag);
    const confirm = window.confirm('Do you really want to add this tag?');
        return Observable
            .of(undefined)
            .filter(() => confirm)
            .mapTo(tag);
  }

**input**

tag5 //which is not in autocompleteitems

**output**
tag parameter is undefined, attached screenshot
![screen shot 2017-05-28 at 8 08 24 am](https://cloud.githubusercontent.com/assets/493559/26529848/1cccf8d8-437d-11e7-967c-f37f323cae3c.png)

**Comment by Gbuomprisco** [comment]
Hi @magarbikram,

There is a bug indeed - I will release a fix soon

**Comment by magarbikram** [comment]
thank you so much @Gbuomprisco , #appreciated

**Comment by magarbikram** [comment]
Just checked a new release which fixed the issue where I was getting null value.

But figured out there is another bug, since I am creating a new tag if input is string instead of object, 
but it adds string instead of newly created object.

for eg: new tag is "tag xx", now I return the observable which returns newly created tag in {id, value} format, but its adding tag xx instead of {id, value} object

```
Observable
            .of(this.getItemDisplay(item))
            .map(display => this.setInputValue(display))
            .filter(display => this.inputForm.form.valid && !!display)
            .map((display: string) => **this.createTag(fromAutocomplete ? item : display))**
            .filter(validationFilter)
            .subscribe(appendItem, undefined, reset);
```

this is where logic got failed.  **this.createTag(fromAutocomplete ? item : display))**

**Comment by magarbikram** [comment]
last release fix this problem, thanks @Gbuomprisco
